src/mapping/query/function/parse_syslog.rs

```diff
@@ -0,0 +1,170 @@
+/// Otherwise, take the current year.
+fn resolve_year((month, _date, _hour, _min, _sec): IncompleteDate) -> i32 {
+    let now = Utc::now();
+    if now.month() == 1 && month == 12 {
+        now.year() - 1
+    } else {
+        now.year()
+    }
+}
+
+/// Create a Value::Map from the fields of the given syslog message.
+fn message_to_value(message: Message<&str>) -> Value {
+    let mut result = BTreeMap::new();
+
+    result.insert("message".to_string(), Value::from(message.msg.to_string()));
+
+    if let Some(host) = message.hostname {
+        result.insert("hostname".to_string(), Value::from(host.to_string()));
+    }
+
+    if let Some(severity) = message.severity {
+        result.insert(
+            "severity".to_string(),
+            Value::from(severity.as_str().to_owned()),
+        );
+    }
+
+    if let Some(facility) = message.facility {
+        result.insert(
+            "facility".to_string(),
+            Value::from(facility.as_str().to_owned()),
+        );
+    }
+
+    if let Some(app_name) = message.appname {
+        result.insert("appname".to_string(), Value::from(app_name.to_owned()));
+    }
+
+    if let Some(msg_id) = message.msgid {
+        result.insert("msgid".to_string(), Value::from(msg_id.to_owned()));
+    }
+
+    if let Some(timestamp) = message.timestamp {
+        let timestamp: DateTime<Utc> = timestamp.into();
+        result.insert("timestamp".to_string(), Value::from(timestamp));
+    }
+
+    if let Some(procid) = message.procid {
+        let value: Value = match procid {
+            ProcId::PID(pid) => pid.into(),
+            ProcId::Name(name) => name.to_string().into(),
+        };
+        result.insert("procid".to_string(), value);
+    }
+
+    for element in message.structured_data.into_iter() {
+        for (name, value) in element.params.into_iter() {
+            let key = format!("{}.{}", element.id, name);
+            result.insert(key, Value::from(value.to_string()));
+        }
+    }
+
+    Value::from(result)
+}
+
+impl Function for ParseSyslogFn {
+    fn execute(&self, ctx: &Event) -> Result<Value> {
+        let message =
+            required!(ctx, self.query, Value::Bytes(v) => String::from_utf8_lossy(&v).into_owned());
+
+        let parsed = syslog_loose::parse_message_with_year(&message, resolve_year);
+
+        Ok(message_to_value(parsed))
+    }
+
+    fn parameters() -> &'static [Parameter] {
+        &[Parameter {
+            keyword: "value",
+            accepts: |v| matches!(v, Value::Bytes(_)),
+            required: true,
+        }]
+    }
+}
+
+impl TryFrom<ArgumentList> for ParseSyslogFn {
+    type Error = String;
+
+    fn try_from(mut arguments: ArgumentList) -> Result<Self> {
+        let query = arguments.required("value")?;
+
+        Ok(Self { query })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::mapping::query::path::Path;
+    use chrono::prelude::*;
+
+    #[test]
+    fn parses() {
```

Comment:
It seems like the `syslog` source has number of tests testing various output permutations. Would it make sense to include them here too? Alternatively, I could see those tests living in `syslog_loose`.

Reply:
I think `syslog_loose` covers the various parsing scenarios, but yes, I think more tests here to test what happens with various fields missing makes a lot of sense.